be/src/runtime/query_context.cpp

```diff
@@ -162,12 +166,13 @@ bool QueryContext::cancel(bool v, std::string msg, Status new_status, int fragme
 
     set_ready_to_execute(true);
     {
-        std::lock_guard<std::mutex> plock(pipeline_lock);
         for (auto& ctx : fragment_id_to_pipeline_ctx) {
             if (fragment_id == ctx.first) {
                 continue;
             }
-            ctx.second->cancel(PPlanFragmentCancelReason::INTERNAL_ERROR, msg);
+            if (auto pipeline_ctx = ctx.second.lock()) {
```

Comment:
we need a lock to protect fragment_id_to_pipeline_ctx.
For example, one thread is exec_plan_fragment, it will use fragment mgr's lock and insert to the map;
and here, the cancel logic will iterate the map, it may core.

And we should not expose the structure fragment_id_to_pipeline_ctx to fragment mgr.